2500× hashrate boost: 20k → up to 50m ops/s (cpu-fast, metrics, throttling, stable)

grenade opened the pull request:
overview
this pr implements the 01–07 implementation docs and delivers a production-ready, observable cpu miner with a clean service boundary, consistent cli, and ci/toolchain hygiene. it intentionally defers select items from the 00 roadmap for post-merge follow-ups.

what’s included (docs/implementation)
- 01: metrics toggle and soc restructure
  - clear separation of concerns across crates (service, engines, pow-core, metrics).
  - optional prometheus exporter via --metrics-port; default off.
- 02: optimized cpu engine and richer metrics
  - new cpu-fast engine based on precompute + incremental modular multiplication (fast path).
  - vendored pow-core with refined api for optimization (future montgomery hook preserved).
  - per-engine/job/thread metrics foundation.
- 03: engine comparability
  - consistent resource settings and terminology for fair, reproducible comparisons.
  - guidance on apples-to-apples testing across engines and instances.
- 04: metrics chunking and dashboards
  - thread-level progress/throughput using chunked subranges with ema smoothing.
  - remove-on-end metrics semantics to avoid scrape artifacts; dashboard composition updated.
- 05: workers and dashboards sync
  - cli clarifies worker semantics (--workers); auto-detects effective cpus honoring cpusets.
  - range partitioning safety (saturating arithmetic, clamped boundaries).
  - test fixes to remove flakiness and align runtime vs. dashboard views.
- 06: throttling engine and difficulty overlay
  - new cpu-chain-manipulator engine that slows incrementally per solved block to help reduce difficulty overshoot on testnets.
  - configurable knobs: --manip-solved-blocks, --manip-base-delay-ns, --manip-step-batch, --manip-throttle-cap.
- 07: stable toolchain, clippy/taplo hygiene, gpu engine placeholders
  - move to stable toolchain (with clippy, rustfmt); remove unused wasm target.
  - clippy cleanup (inline format args, doc lint fixes) and taplo formatting; ci-ready.
  - keep cpu-prefixed cli engine variants; add gpu-cuda/gpu-opencl placeholders.
  - selecting gpu engines errs early with a clear, user-facing message.

cli highlights
- engine selection: --engine cpu-fast (default), cpu-baseline, cpu-chain-manipulator; gpu-cuda|gpu-opencl return a clear “unimplemented” error.
- resource control: --workers <n> (respects cgroups cpuset); balanced default when unspecified.
- observability: --metrics-port <port> to enable prometheus exporter.
- metrics cadence: --progress-chunk-ms <ms> to tune per-thread update frequency.
- throttling controls for cpu-chain-manipulator: --manip-solved-blocks, --manip-base-delay-ns, --manip-step-batch, --manip-throttle-cap.

validation
- cargo +stable test --locked --workspace: green.
- cargo clippy --workspace --all-targets (clean).
- taplo fmt --check (clean).
- manual runtime sanity across engines with and without metrics enabled.

deferred (from 00 roadmap; to follow after merge)
- montgomery multiplication/accelerated modular arithmetic in pow-core.
- gpu offloading engines (cuda/opencl) and corresponding metrics/dashboards.
- additional integration/e2e coverage and advanced dashboard alerting/recording rules.

reviewer guide
- full iteration docs: docs/implementation/readme.md
  - 07-prompt/response: stable toolchain, clippy/taplo, gpu enum placeholders
  - 06-prompt/response: manipulator engine and difficulty overlay
  - 05-prompt/response: workers and dashboards sync
  - 04-prompt/response: metrics chunking and dashboards
  - 03-prompt/response: engine comparability
  - 02-prompt/response: cpu-fast engine and metrics
  - 01-prompt/response: metrics toggle and soc restructure

n13 reviewed (approved):
IMO anything goes in the external miner, since we don't really care how it arrives at good results. 

So there's duplicate code to the node, but it's OK. 

Protocol was not changed.

n13 commented on crates/engine-gpu-cuda/src/lib.rs:
```diff
@@ -0,0 +1,99 @@
+#![deny(rust_2018_idioms)]
```
If this code is not used, let's remove it and put it in a branch for the time being. Dead code has a habit of accumulating.

grenade replied:
see: https://github.com/Quantus-Network/quantus-miner/pull/1#issuecomment-3273047364

n13 reviewed (changes requested):
Ok requesting this
- remove dead code
- run through clippy
- remove "explanations" which are not neded the whole /implementation segment

Other than that, Dastan pointed out there's many things that could be done in much fewer lines of code but I don't think it's worth our time rewriting what AI wrote. 

So just requesting the above.

grenade commented:
- clippy: was run many times during implementation and it runs on every push thanks to ci.yml. i'm struggling to understand why it was flagged twice since it is really obvious in ci, that it is being run and passing. i must assume that this was overlooked.
- the “dead code” feedback, appears to be about placeholder code (e.g., gpu placeholders). gpu placeholders are used intentionally for ux and erroring early with a clear message; they are exercised by tests and keep cli consistent. i did not remove them because they are intentional placeholders with a defined behavior and tests. they are the next order of business for implementation once these changes land, so ripping them out now to satisfy a whim is busy-work-for-bureaucrats twice, since i would be putting them back in immediately following the merge. they exist intentionally and support the next batch of changes. i'm not inclined to jump through this hoop since it creates busy work for me with no upside.
- logging and serviceconfig: good suggestion. i've implemented display for serviceconfig and switched cli startup to log a concise ‘effective config’ line. the miner-service also logs the final resolved configuration after engine selection.
- usize comparison note: agreed. i removed that useless comparison in tests. there’s no remaining usize-negative comparison in the updated code.
- implementation docs: removed.

grenade commented:
<img width="1812" height="990" alt="image" src="https://github.com/user-attachments/assets/d047cb58-8e81-4b9f-a64e-b858b8dee78b" />
for reference:
- the blue line at 0 ops/s in *Global Hash Rate by Instance* represents miners with the old 20k hashrate on resonance
- the hashrates in the 20m to 40m range represent miners running the optimisations in this pr. on heisenberg

n13 commented:
Approved
